File: leetcode-and-interview-questions/singleNumber/singleNumber.js
// Given a non-empty array of integers nums, every element appears twice except for one. Find that single one.
// You must implement a solution with a linear runtime complexity and use only constant extra space.

// Input: nums = [2,2,1]
// Output: 1

// Input: nums = [4,1,2,1,2]
// Output: 4

// Input: nums = [1]
// Output: 1

/**
 * @param {number[]} nums
 * @return {number}
 */

// O(2N)
var singleNumber = function(nums) {
  let seen = {}
  nums.forEach(num => {
    if (seen[num]) seen[num] += 1;
    else seen[num] = 1;
  })
  for (let n in seen) {
    if (seen[n] === 1) return n;
  }
  return null;
};

// O(N) solution
var singleNumber = function(nums) {
  let res = 0;
  for (let i = 0; i < nums.length; i++) {
    res ^= nums[i];
  }
  return res;
}

console.log(singleNumber([2, 2, 1]))
console.log(singleNumber([4,1,2,1,2]))

File: leetcode-and-interview-questions/singleNumber/singleNumber.py
import functools


# My solution
class Solution:
  def singleNumber(self, nums: List[int]) -> int:
    seen = {}
    for n in nums:
      if n in seen:
        seen[n] += 1
      else:
        seen[n] = 1
    for i in seen:
      if seen[i] == 1:
        return i
    return None


# Sleeker solution
class Solution:
  def singleNumber(self, nums):
    seen = {}
    for n in nums:
      seen[n] = seen.get(n, 0) + 1
    for k, v in nums.items():
      if v == 1:
        return k


# O(N) solution
class Solution:
  def singleNumber(self, nums):
    res = 0
    for n in nums:
      res ^= n
    return res


# One liner
class Solution:
  def singleNumber(self, nums):
    return reduce(lambda x, y: x ^ y, nums)

class Solution:
  def singleNumber(self, nums):
    for i in range(1,len(nums)):
        nums[0] ^= nums[i]
    return nums[0]